test-data/unit/check-overloading.test

```diff
@@ -1136,3 +1136,24 @@ def f(a: Callable[[], int]) -> None: pass
 @overload
 def f(a: str) -> None: pass
 f(0)  # E: No overload variant of "f" matches argument types [builtins.int]
+
+[case testCustomRedefinitionDecorator]
+from typing import Any, Callable, Type
+
+class Chain(object):
+    def chain(self, function: Callable[[Any], int]) -> 'Chain':
+        return self
+
+class Test(object):
+    do_chain = Chain()
+
+    @do_chain.chain
+    def do_chain(self) -> int:
+        return 2
+
+    @do_chain.chain  # E: Name 'do_chain' already defined
```

Comment:
I think deciding that what you meant here was ok is out of scope.  `type: ignore` might be what you want here.